ha/lock.go

```diff
@@ -0,0 +1,127 @@
+package ha
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/uuid"
+)
+
+const (
+	renewalFactor = 0.7
+	waitFactor    = 1.1
+)
+
+type lock interface {
+	Acquire(ctx context.Context, callerID string) (string, error)
+	Release(ctx context.Context) error
+	Renew(ctx context.Context) error
+}
+
+type HAController struct {
+	ID            string
+	renewalPeriod time.Duration
+	waitPeriod    time.Duration
+	randomDelay   time.Duration
+
+	logger log.Logger
+	lock   lock
+}
+
+func NewHAController(l lock, logger log.Logger, lease time.Duration) *HAController {
+	logger = logger.Named("ha_mode")
+
+	rn := rand.New(rand.NewSource(time.Now().Unix())).Intn(100)
+	hac := HAController{
+		lock:          l,
+		logger:        logger,
+		renewalPeriod: time.Duration(float64(lease) * renewalFactor),
+		waitPeriod:    time.Duration(float64(lease) * waitFactor),
+		ID:            uuid.Generate(),
+		randomDelay:   time.Duration(rn) * time.Millisecond,
+	}
+
+	return &hac
+}
+
+func (hc *HAController) Start(ctx context.Context, protectedFunc func(ctx context.Context)) error {
+	hc.logger.Named(hc.ID)
+
+	// To avoid collisions if all the instances start at the same time, wait
+	// a random time before making the first call.
+	hc.wait(ctx)
+
+	waitTicker := time.NewTicker(hc.waitPeriod)
+	defer waitTicker.Stop()
+
+	for {
+		hc.logger.Debug("attempting to acquire lock")
+		lockID, err := hc.lock.Acquire(ctx, hc.ID)
+		if err != nil {
+			// TODO: What to do with fatal errors?
+			hc.logger.Error("unable to get lock", err)
+		}
+
+		if lockID != "" {
+			hc.logger.Debug("lock acquired, ID", lockID)
+			funcCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// Start running the lock protected function
+			go protectedFunc(funcCtx)
+
+			// Maintain lease is a blocking function, will only return in case
+			// the lock is lost or the context is canceled.
+			err := hc.maintainLease(ctx)
+			if err != nil {
+				hc.logger.Debug("lease lost", err)
+				cancel()
+				// Give the protected function some time to return before potentially
+				// running it again.
+				hc.wait(ctx)
+			}
+		}
+
+		waitTicker = time.NewTicker(hc.waitPeriod)
```

Comment:
I tried that one, but the waitTicker.C channel is buffered and if the a tick has already gone before, we would have had to drain it, just like the timer, and in the case of an instance that was previously holding the lock, this is 100% the case, the options were to leave it as it was or to create a new ticker, like I did. Im not aware of any possible memory problems this can bring, but if you do, please tell me

Reply:
I forgot to stop the ticker before rewriting it, updated it

Reply:
LGTM!

I had to check for myself because I'm paranoid about this sort of thing and https://go.dev/play/p/kdMH0bva-5Z and https://go.dev/play/p/EV6J--iRBLf are equivalent in terms of the resulting stack traces; neither leaks.

Reply:
Double checks are always welcomed!